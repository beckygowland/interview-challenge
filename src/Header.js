import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

class Header extends React.PureComponent {
    render() {
        const { count, summary } = this.props;
        return (
            <div className="menu-summary">
                <div className="container">
                    <div className="row">
                        <div className="col-6 menu-summary-left">
                            <span>{count} items</span>
                        </div>
                        <div className="col-6 menu-summary-right">
                            {Object.keys(summary).map((sumItem) => (
                                <span key={sumItem}>{summary[sumItem]}x
                                    <span className="dietary">{sumItem}</span>
                                </span>
                            ))}
                        </div>
                    </div>
                </div>
            </div>
        )
    };
}

Header.propTypes = {
    count: PropTypes.number.isRequired,
    summary: PropTypes.object.isRequired,
};

function getDietarySummary(foodItems) {
    let breakdown = {};
    for (let item of foodItems) {
        for (let dietaryItem of item.dietaries) {
            breakdown[dietaryItem] = breakdown[dietaryItem] !== undefined ?
                breakdown[dietaryItem] + 1 : 1;
        }
    }
    return breakdown;
}

const mapStateToProps = ({ selectedItems }) => ({
    count: selectedItems.length,
    summary: getDietarySummary(selectedItems)
});

export default connect(mapStateToProps, {})(Header);