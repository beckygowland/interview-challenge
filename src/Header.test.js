import React from 'react';
import { Header , mapStateToProps } from './Header';

import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

it('renders', () => {
  const div = shallow(<Header count={0} summary={{}} />);
  expect(div).toMatchSnapshot();
  expect(div.find('#count').text()).toBe('0 items');
  expect(div.find('.dietary').length).toBe(0);
});

it('set proper count and dietary breakdown', () => {
  const summary = {a: 3, b: 2};
  const div = shallow(<Header count={3} summary={summary} />);
  
  expect(div.find('#count').text()).toBe('3 items');
  const summaryItems = div.find('.menu-summary-right'); 
  expect(summaryItems.children().length).toBe(2);
  summaryItems.children().forEach((node, i) => {
    const key = Object.keys(summary)[i];
    const count = summary[key] + 'x';
    expect(node.text()).toBe(count + key);
  })
});

it('mapStateToProps', () => {
  const selectedItems = [
    { id: 1, dietaries: ['a', 'b', 'c'] },
    { id: 2, dietaries: ['b', 'c'] },
    { id: 1, dietaries: ['c', 'd'] },
  ]
  const props = mapStateToProps({selectedItems})
  expect(props.count).toBe(3);
  expect(props.summary.a).toBe(1);
  expect(props.summary.b).toBe(2);
  expect(props.summary.c).toBe(3);
  expect(props.summary.d).toBe(1);
});
