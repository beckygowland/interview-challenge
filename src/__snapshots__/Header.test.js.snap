// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Header
    count={0}
    summary={Object {}}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="container"
>
        <div
                className="row"
        >
                <div
                        className="col-6 menu-summary-left"
                >
                        <span
                                id="count"
                        >
                                0
                                 items
                        </span>
                </div>
                <div
                        className="col-6 menu-summary-right"
                >
                        
                </div>
        </div>
</div>,
      "className": "menu-summary",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="row"
>
          <div
                    className="col-6 menu-summary-left"
          >
                    <span
                              id="count"
                    >
                              0
                               items
                    </span>
          </div>
          <div
                    className="col-6 menu-summary-right"
          >
                    
          </div>
</div>,
        "className": "container",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="col-6 menu-summary-left"
>
              <span
                            id="count"
              >
                            0
                             items
              </span>
</div>,
            <div
              className="col-6 menu-summary-right"
>
              
</div>,
          ],
          "className": "row",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <span
                id="count"
>
                0
                 items
</span>,
              "className": "col-6 menu-summary-left",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  0,
                  " items",
                ],
                "id": "count",
              },
              "ref": null,
              "rendered": Array [
                0,
                " items",
              ],
              "type": "span",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [],
              "className": "col-6 menu-summary-right",
            },
            "ref": null,
            "rendered": Array [],
            "type": "div",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="container"
>
          <div
                    className="row"
          >
                    <div
                              className="col-6 menu-summary-left"
                    >
                              <span
                                        id="count"
                              >
                                        0
                                         items
                              </span>
                    </div>
                    <div
                              className="col-6 menu-summary-right"
                    >
                              
                    </div>
          </div>
</div>,
        "className": "menu-summary",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="row"
>
            <div
                        className="col-6 menu-summary-left"
            >
                        <span
                                    id="count"
                        >
                                    0
                                     items
                        </span>
            </div>
            <div
                        className="col-6 menu-summary-right"
            >
                        
            </div>
</div>,
          "className": "container",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="col-6 menu-summary-left"
>
                <span
                                id="count"
                >
                                0
                                 items
                </span>
</div>,
              <div
                className="col-6 menu-summary-right"
>
                
</div>,
            ],
            "className": "row",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <span
                  id="count"
>
                  0
                   items
</span>,
                "className": "col-6 menu-summary-left",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    0,
                    " items",
                  ],
                  "id": "count",
                },
                "ref": null,
                "rendered": Array [
                  0,
                  " items",
                ],
                "type": "span",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [],
                "className": "col-6 menu-summary-right",
              },
              "ref": null,
              "rendered": Array [],
              "type": "div",
            },
          ],
          "type": "div",
        },
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
